src/main/telemetry/msp_shared.c

```diff
@@ -29,225 +29,297 @@
+            requestPacket.result = 0;
+            requestPacket.buf.ptr = requestBuffer;
+            requestPacket.buf.end = requestBuffer + mspPayloadSize;
+            mspStarted = 1;
+        } else { // this MSP packet is too big to fit in the buffer.
+            sendMspErrorResponse(TELEMETRY_MSP_REQUEST_IS_TOO_BIG, requestPacket.cmd);
+            return true;
+        }
+    } else { // second onward chunk
+        if (!mspStarted) { // no start packet yet, throw this one away
+            return false;
+        } else {
+            if (((lastSeq + 1) & MSP_STATUS_SEQUENCE_MASK) != seqNumber) {
+                // packet loss detected!
+                mspStarted = 0;
+                return false;
+            }
+        }
+        sbufInit(&sbufInput, crsfPayload +1, crsfPayload + crsfPayloadLength);
     }
 
-    const uint8_t bufferBytesRemaining = sbufBytesRemaining(rxBuf);
-    const uint8_t frameBytesRemaining = sbufBytesRemaining(frameBuf);
-    uint8_t payload[frameBytesRemaining];
+    lastSeq = seqNumber;
 
-    if (bufferBytesRemaining >= frameBytesRemaining) {
-        sbufReadData(frameBuf, payload, frameBytesRemaining);
-        sbufAdvance(frameBuf, frameBytesRemaining);
-        sbufWriteData(rxBuf, payload, frameBytesRemaining);
-        lastSeq = seqNumber;
+    const int payloadExpecting = sbufBytesRemaining(&requestPacket.buf);
+    const int payloadIncoming = sbufBytesRemaining(&sbufInput);
 
+    if (payloadExpecting > payloadIncoming) {
+        sbufWriteData(&requestPacket.buf, sbufInput.ptr, payloadIncoming);
+        sbufAdvance(&sbufInput, payloadIncoming);
         return false;
-    } else {
-        sbufReadData(frameBuf, payload, bufferBytesRemaining);
-        sbufAdvance(frameBuf, bufferBytesRemaining);
-        sbufWriteData(rxBuf, payload, bufferBytesRemaining);
-        sbufSwitchToReader(rxBuf, mspPackage.requestBuffer);
-        while (sbufBytesRemaining(rxBuf)) {
-            checksum ^= sbufReadU8(rxBuf);
-        }
-
-        if (checksum != *frameBuf->ptr) {
-            mspStarted = 0;
-            sendMspErrorResponse(TELEMETRY_MSP_CRC_ERROR, packet->cmd);
-            return true;
+    } else { // this is the last/only chunk 
+        if (payloadExpecting) {
+            sbufWriteData(&requestPacket.buf, sbufInput.ptr, payloadExpecting);
+            sbufAdvance(&sbufInput, payloadExpecting);
         }
+        // the whole frame is received
+        sbufSwitchToReader(&requestPacket.buf, requestBuffer);
     }
 
     // Skip a few telemetry requests if command is MSP_EEPROM_WRITE
-    if (packet->cmd == MSP_EEPROM_WRITE && skipsBeforeResponse) {
+    if (requestPacket.cmd == MSP_EEPROM_WRITE && skipsBeforeResponse) {
         *skipsBeforeResponse = TELEMETRY_REQUEST_SKIPS_AFTER_EEPROMWRITE;
     }
 
     mspStarted = 0;
-    sbufSwitchToReader(rxBuf, mspPackage.requestBuffer);
+    sbufSwitchToReader(&requestPacket.buf, requestBuffer);
     processMspPacket();
     return true;
```

Comment:
```sbufSwitchToReader()``` is called twice here. This is redundant and won't work properly. We'll end up with ```requestPacket.buf.ptr == requestPacket.buf.end``` and lose the data. A single call at the end is enough.

Reply:
wow, that's huge! My bad. Thanks for noticing.